application/src/main/java/org/opentripplanner/osm/tagmapping/OsmTagMapper.java

```diff
@@ -1,18 +1,728 @@
 package org.opentripplanner.osm.tagmapping;
 
+import static org.opentripplanner.osm.wayproperty.MixinPropertiesBuilder.ofBicycleSafety;
+import static org.opentripplanner.osm.wayproperty.MixinPropertiesBuilder.ofWalkSafety;
+import static org.opentripplanner.osm.wayproperty.WayPropertiesBuilder.withModes;
+import static org.opentripplanner.street.model.StreetTraversalPermission.ALL;
+import static org.opentripplanner.street.model.StreetTraversalPermission.BICYCLE_AND_CAR;
+import static org.opentripplanner.street.model.StreetTraversalPermission.CAR;
+import static org.opentripplanner.street.model.StreetTraversalPermission.NONE;
+import static org.opentripplanner.street.model.StreetTraversalPermission.PEDESTRIAN;
+import static org.opentripplanner.street.model.StreetTraversalPermission.PEDESTRIAN_AND_BICYCLE;
+
 import org.opentripplanner.osm.model.OsmWithTags;
+import org.opentripplanner.osm.wayproperty.WayProperties;
 import org.opentripplanner.osm.wayproperty.WayPropertySet;
+import org.opentripplanner.osm.wayproperty.specifier.BestMatchSpecifier;
+import org.opentripplanner.osm.wayproperty.specifier.LogicalOrSpecifier;
+import org.opentripplanner.routing.services.notes.StreetNotesService;
 
 /**
- * Interface for populating a {@link WayPropertySet} that determine how OSM streets can be traversed
- * in various modes and named.
+ * This factory class provides a default collection of {@link WayProperties} that determine how OSM
+ * streets can be traversed in various modes.
+ * <p>
+ * Circa January 2011, Grant and Mele at TriMet undertook proper testing of bike (and transit)
+ * routing, and worked with David Turner on assigning proper weights to different facility types.
+ * The weights in this file grew organically from trial and error, and are the result of months of
+ * testing and tweaking the routes that OTP returned, as well as actually walking/biking these
+ * routes and making changes based on those experiences. This set of weights should be a great
+ * starting point for others to use, but they are to some extent tailored to the situation in
+ * Portland and people shouldn't hesitate to adjust them to for their own instance.
+ * <p>
+ * The rules for assigning WayProperties to OSM ways are explained in. The final tie breaker if two
+ * Pickers both match is the sequence that the properties are added in this file: if all else is
+ * equal the 'props.setProperties' statement that is closer to the top of the page will prevail over
+ * those lower down the page.
+ * <p>
+ * Foot and bicycle permissions are also addressed in OpenStreetMapGraphBuilderImpl.Handler#getPermissionsForEntity().
+ * For instance, if a way that normally does not permit walking based on its tag matches (the
+ * prevailing 'props.setProperties' statement) has a 'foot=yes' tag the permissions are overridden
+ * and walking is allowed on that way.
+ * <p>
+ * TODO clarify why this needs a separate factory interface.
```

Comment:
First I don´t know this code well. But, inheritance is usually a bad chose if all you want to do is to share code - or even worse, save typing on the keyboard. 

Interfaces are a contract between the caller and the implementation. Interfaces should be used if you want to document the API needed to plug into something else. In this case the contract is:
- `populateProperties(WayPropertySet wayPropertySet)` - used by `OsmProvider`
- `getCarSpeedForWay(OsmWithTags way, boolean backward)` - used by `OsmModule and WalkableAreaBuilder`
- `getMaxUsedCarSpeed(WayPropertySet wayPropertySet)` - used by `OsmModule`
- `boolean isMotorVehicleThroughTrafficExplicitlyDisallowed(OsmWithTags way)` used by `SafetyValueNormalizer`
- `boolean isBicycleNoThroughTrafficExplicitlyDisallowed(OsmWithTags way)` used by `SafetyValueNormalizer`

I am not sure if OsmTagMapper play the _same_ role in `OsmProvider`, `OsmModule`,`WalkableAreaBuilder`, `SafetyValueNormalizer`. But if so an interface with the methods above would be nice. There is no overlap in the usage by `OsmProvider`, `OsmModule` and `WalkableAreaBuilder` so maybe zero, two or even three interfaces is the right thing. 

**What about the other methods?**
 - boolean doesTagValueDisallowThroughTraffic(String tagValue)
 - boolean isGeneralNoThroughTraffic(OsmWithTags way)
 - boolean isVehicleThroughTrafficExplicitlyDisallowed(OsmWithTags way)

These methods are abstract factory methods (I think). So, they do not belong in the contract - and they should be package local. 

I looked at the overridden methods as well, but quickly fall off. One alternative to factory methods is the strategy pattern. I do not know what is the best solution in this case. 

I would call this class the `DefaultOsmTagMapper`, extract a `OsmTagMapper` interface with 5 methods in it.

Reply:
I would think that a regular java class with its public methods defines equally good contract for the caller. Am I  wrong?

The changes I made to the tag mapper classes clearly reduce code lines and  also make the responsibilities obvious. Previously, some methods were inherited to interface, some were delegated to the default tag mapper. This kind of complication should be avoided.

Reply:
A regular base Java class is not as clear. Some times you need to add public methods for other reasons and the contract get cluttered with other stuff, like the implementation of the contract and fields/other methods. But in this case I think it is ok - it depends on what the "bounded context" is. When you cross boarder you want clear contract, while it it is less important inside a module. To me adding on the `getCarSpeedForWay(OsmWithTags way, boolean backward)`, and `getMaxUsedCarSpeed(WayPropertySet wayPropertySet)` seams like a hack to solve the constant-speed problem. The OsmTagMapper does not have a single-responsibility. Maybe we can spend 10 minutes on this in the dev meeting today.

I agree that the changes you did is a good improvement 😃